vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,666 @@
+  //   data.include_list.insert(included_action_id);
+
+  // set the available actions in the cb multi-example
+  uint32_t index = 0;
+  data.origin_index.clear();
+  data.origin_index.resize(data.actions.size(), 0);
+  for (size_t i = 0; i < data.actions.size(); i++)
+  {
+    // For V0, include list is not supported
+    // filter actions that are not explicitly included
+    // if (!data.include_list.empty() && data.include_list.find((uint32_t)i) == data.include_list.end())
+    //   continue;
+
+    // filter actions chosen by previous slots
+    if (data.exclude_list[i])
+      continue;
+
+    // select the action
+    cb_ex.push_back(data.actions[i]);
+
+    // save the original index from the root multi-example
+    data.origin_index[index++] = (uint32_t)i;
+
+    // remember the index of the chosen action
+    if (is_learn && slot_has_label && i == slot->l.conditional_contextual_bandit.outcome->probabilities[0].action)
+    {
+      // This is used to remove the label later.
+      data.action_with_label = (uint32_t)i;
+      attach_label_to_example(index, data.actions[i], slot->l.conditional_contextual_bandit.outcome, data);
+    }
+  }
+
+  // Must reset this in case the pooled example has stale data here.
+  data.shared->pred.a_s = data.action_score_pool.get_object();
+
+  // Tag can be used for specifying the sampling seed per slot. For it to be used it must be inserted into the shared
+  // example.
+  std::swap(data.shared->tag, slot->tag);
+}
+
+// iterate over slots contained in the multi-example, and for each slot, build a cb example and perform a
+// cb_explore_adf call.
+template <bool is_learn>
+void learn_or_predict(ccb& data, multi_learner& base, multi_ex& examples)
+{
+  clear_all(data);
+  split_multi_example(examples, data);  // split shared, actions and slots
+
+#ifndef NDEBUG
+  sanity_checks<is_learn>(data);
+#endif
+
+  // Stash the CCB labels before rewriting them.
+  data.stored_labels.clear();
+  for (auto ex : examples)
+  {
+    data.stored_labels.push_back({ex->l.conditional_contextual_bandit.type, ex->l.conditional_contextual_bandit.outcome,
+        ex->l.conditional_contextual_bandit.explicit_included_actions});
+  }
+
+  // This will overwrite the labels with CB.
+  create_cb_labels(data);
+
+  // Reset exclusion list for this example.
+  data.exclude_list.assign(data.actions.size(), false);
+
+  auto decision_scores = data.action_scores_pool.get_object();
+
+  // for each slot, re-build the cb example and call cb_explore_adf
+  int slot_id = 0;
+  for (example* slot : data.slots)
+  {
+    // Namespace crossing for slot features.
+    // If the slot example only has the constant namespace, there will be no extra crossing and so skip that logic.
+    if (!(slot->indices.size() == 1 && slot->indices[0] == constant_namespace))
+    {
+      data.generated_interactions.clear();
+      std::copy(data.original_interactions->begin(), data.original_interactions->end(),
```

Comment:
Use vector assignment operator?  Easier to understand.  Might have vector specific optimizations depending on impl.  (might alloc memory at the beginning, for example)

Reply:
I moved to using a set due to duplicate issues, assignment can no longer be used.